MeshAlgo, ShapeAlgo, ParameterAlgo: support Pref in Arnold

mattigruener opened the pull request:
In order to support Pref for Arnold, we need to remove the 'user:'
prefix and be able to distinguish between vector and point types.

Hey @johnhaddon, Is this roughly what you had in mind when we talked about it in Shotgun? I'll add the 'onHold' label because I should probably change a whole bunch of tests that seem to use the prefix.

mattigruener commented:
I wanted to also add the warning for clashing parameters, but didn't get to finish it because of a handful of other things that came up. Will add that tomorrow.

johnhaddon commented on contrib/IECoreArnold/include/IECoreArnold/ParameterAlgo.h:
```diff
@@ -62,6 +62,7 @@ IECOREARNOLD_API void getParameters( AtNode *node, IECore::CompoundDataMap &valu
 /// or false depending on whether or not the Arnold type will be an
 /// array. Returns AI_TYPE_NONE if there is no suitable Arnold type.
 IECOREARNOLD_API int parameterType( IECore::TypeId dataType, bool &array );
+IECOREARNOLD_API int parameterType( IECore::DataPtr data, bool &array );
```
Should be `const IECore::Data *data`, since we don't need to modify the data or own a reference to it.

johnhaddon commented:
Thanks Matti - this is exactly what I had in mind.

mattigruener commented:
@johnhaddon I added the warning in two spots because our handling of meshes vs everything else is a little different at the moment. Maybe we want to unify that at some point? I also added a few built-ins, but I guess we would want to add a whole lot more here? This is a list of names we should probably look for:

> kick -info polymesh

node:         polymesh
type:         shape
output:       (null)
parameters:   44
filename:     <built-in>
version:      4.2.14.2.wip.osl2

Type          Name                              Default
------------  --------------------------------  --------------------------------
UINT[]        nsides                            (empty)
UINT[]        vidxs                             (empty)
UINT[]        nidxs                             (empty)
UINT[]        uvidxs                            (empty)
UINT[]        crease_idxs                       (empty)
FLOAT[]       crease_sharpness                  (empty)
BYTE[]        shidxs                            (empty)
POINT[]       vlist                             (empty)
VECTOR[]      nlist                             (empty)
POINT2[]      uvlist                            (empty)
BOOL          smoothing                         false
ENUM          subdiv_type                       none
BYTE          subdiv_iterations                 1
FLOAT         subdiv_adaptive_error             0
NODE          subdiv_dicing_camera              (null)
ENUM          subdiv_adaptive_metric            auto
ENUM          subdiv_adaptive_space             raster
ENUM          subdiv_uv_smoothing               pin_corners
BOOL          subdiv_smooth_derivs              false
NODE[]        disp_map                          (empty)
FLOAT         disp_padding                      0
FLOAT         disp_height                       1
FLOAT         disp_zero_value                   0
BOOL          disp_autobump                     false
BYTE          autobump_visibility               159
BYTE          visibility                        255
BYTE          sidedness                         255
BOOL          receive_shadows                   true
BOOL          self_shadows                      true
BOOL          invert_normals                    false
FLOAT         ray_bias                          1e-06
MATRIX[]      matrix                            (empty)
NODE[]        shader                            (empty)
BOOL          opaque                            true
BOOL          matte                             false
BOOL          use_light_group                   false
NODE[]        light_group                       (empty)
BOOL          use_shadow_group                  false
NODE[]        shadow_group                      (empty)
STRING[]      trace_sets                        (empty)
FLOAT[]       transform_time_samples            (2 elements)
FLOAT[]       deform_time_samples               (2 elements)
INT           id                                0
STRING        name

Transferring this list into the code is a little ugly. Do you have an opinion on how to determine which variables we'd want to ignore?

I also addressed your previous feedback. Thanks :)

johnhaddon commented:
> Transferring this list into the code is a little ugly. Do you have an opinion on how to determine which variables we'd want to ignore?

I don't think we need a hardcoded list at all. Instead we could use `AtiNodeGetNodeEntry()` followed by `AiNodeEntryLookUpParameter()` to determine at runtime if the parameter exists as a build-in. I reckon we could just do this in `convertPrimitiveVariable()`, and return early if there's a clash. This sort of introspection of available nodes and parameters is one of the nicest things about the Arnold API...

mattigruener commented:
That's actually really nice - didn't know the API allows that. I updated the PR. It's much cleaner now. This should do the trick?

I also noticed that the tests that I mentioned before don't really exist and all tests that do some 'user:' stuff are non-Arnold ones.

johnhaddon commented on contrib/IECoreArnold/src/IECoreArnold/ShapeAlgo.cpp:
```diff
@@ -176,6 +176,17 @@ void convertRadius( const std::vector<const IECore::Primitive *> &samples, AtNod
+	if ( AiNodeEntryLookUpParameter(	entry, name ) != NULL ){
+		msg(
+			Msg::Warning,
+			"ToArnoldShapeConverter::convertPrimitiveVariable",
```
`ToArnoldShapeConverter` isn't a think any more (although it once was). I suspect there are lingering references to it in other messages, and you picked this up from one of them? Can we replace with `ShapeAlgo::convertPrimitiveVariable`?

mattigruener replied:
Yeah, I copied that from somewhere else and thought it might be a convention I didn't know about yet ;) I will replace all references to `ToArnoldShapeConverter`.

johnhaddon commented:
Looks good to me - thanks Matti. Since you've discovered that the existing unit tests are weak in this area, how about beefing them up to cover the features you've added?

mattigruener commented:
I updated the tests and changed the log message.

johnhaddon commented:
Nice one - thanks Matti!
